keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,657 @@
+section, we disallow borrowing in the cohort in the current cycle after
+evaluating a Workload that doesn't require borrowing. This is the same behavior
+that we have today, but note that this criteria includes Workloads that need
+preemption, because there is no preemption with borrowing quota.
+
+This guarantees that, in future cycles, we can admit Workloads that were not
+heads in their ClusterQueues in this cycle, but could fit without borrowing in
+the next cycle, before lending quota to other ClusterQueues.
+
+In the past, we only disallowed borrowing in the cohort if we were able to
+admit the Workload, because we only kept track of flavor assignments of type 0.
+This caused ClusterQueues in the cohort to continue
+borrowing quota, even if there were pending Workloads that would fit under the
+min Quota for their ClusterQueues.
+
+It is actually possible to limit borrowing only for the involved flavors,
+instead of restricting borrowing for all the flavors in the cohort. But we will
+leave this as a future possible optimization to improve throughput.
+
+#### Preemption
+
+For each Workload that got flavor assignments of type 1 or 2, we might need to
+preempt some admitted Workloads.
+
+The algorithm goes like follows:
+
+0. Check that `ClusterQueue.spec.waitBeforePreemptionSeconds` is satisfied.
+   The time is based on the `lastTransitionTime` of the condition `Admitted`
+   (with `status=False`). If it doesn't exist, we use the current time.
+   If the period is not satisfied, we abort preemption.
+   
+   If this is a ClusterQueue with BestEffortFIFO, the Workload is put in the
+   inadmissible staging area with a deadline equal to the remaining time to
+   fullfil `waitBeforePreemptionSeconds`.
+
+   If this is a ClusterQueue with StrictFIFO, the Workload will be requeued
+   and retried immediately in the next admission cycle.
+
+1. Check whether preemption could help.
+
+   Preemption is limited to Workloads with a priority lower than the incoming
+   Workload. To avoid unnecessary preemption calculations, we can keep a
+   priority queue with the priorities of active Workloads in the ClusterQueue
+   and the cohort. If the lowest priority in these sets is higher than or equal
+   to the priority of the incoming Workload, we can skip the preemption
+   algorithm.
+
+2. Obtain a list of victim Workloads to be preempted.
+
+  1. If the workload has flavor assignments of type 1, add the Workloads from
+     other ClusterQueues in the cohort that are over their min quota.
+  2. If the workload has flavor assigments of type 2, add the Workloads from
+     the same ClusterQueue.
+
+  When going over these sets, we filter out the Workloads that have a priority
+  greather than or equal to the priority of the incoming Workload or that are
+  not using the flavors that were selected for the incoming Workload.
+  
+  If the list is empty, abort preemption.
+
+3. Sort the Workloads using the following criteria:
+   1. Lower priority first.
```

Comment:
if type 1, shouldn't we preempt workloads from the same CQ last irrespective of priority?

Reply:
TBD from the other thread.